compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -217,6 +240,188 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
+   *   - when encountering an application C[Ts], where C is a type variable or parameter
+   *     that has an instantiation in this constraint, assume the type parameters of
+   *     the instantiation instead of the type parameters of C when traversing the
+   *     arguments Ts. That can make a difference for the variance in which an argument
+   *     is traversed. Example constraint:
+   *
+   *         constrainded types: C[X], A
+   *         A >: C[B]
+   *         C := Option
+   *
+   *     Here, B is traversed with variance +1 instead of 0. Test case: pos/t3152.scala
+   *
+   *   - When typing a prefx, don't avoid negative variances. This matters only for the
+   *     corner case where a parameter is instantiated to Nothing (see comment in
+   *     TypeAccumulator#applyToPrefix). When determining instantiation directions in
+   *     interpolations (which is what dependency variances are for), it can be ignored.
+   */
+  private trait ConstraintAwareTraversal[T] extends TypeAccumulator[T]:
+
+    override def tyconTypeParams(tp: AppliedType)(using Context): List[ParamInfo] =
+      def tparams(tycon: Type): List[ParamInfo] = tycon match
+        case tycon: TypeVar if !tycon.isInstantiated => tparams(tycon.origin)
+        case tycon: TypeParamRef =>
+          entry(tycon) match
+            case _: TypeBounds => tp.tyconTypeParams
+            case tycon1 if tycon1.typeParams.nonEmpty => tycon1.typeParams
+            case _ => tp.tyconTypeParams
+        case _ => tp.tyconTypeParams
+      tparams(tp.tycon)
+
+    override def applyToPrefix(x: T, tp: NamedType): T =
+      this(x, tp.prefix)
+  end ConstraintAwareTraversal
+
+  private class Adjuster(srcParam: TypeParamRef)(using Context)
+  extends TypeTraverser, ConstraintAwareTraversal[Unit]:
+
+    var add: Boolean = compiletime.uninitialized
+    private val seen = util.HashSet[LazyRef]()
+
+    def update(deps: ReverseDeps, referenced: TypeParamRef): ReverseDeps =
+      val prev = deps.at(referenced)
+      val newSet = if add then prev + srcParam else prev - srcParam
+      if newSet.isEmpty then deps.remove(referenced)
+      else deps.updated(referenced, newSet)
+
+    def traverse(t: Type) = t match
+      case param: TypeParamRef =>
+        entry(param) match
+          case _: TypeBounds =>
+            if variance >= 0 then coDeps = update(coDeps, param)
+            if variance <= 0 then contraDeps = update(contraDeps, param)
+          case tp =>
+            traverse(tp)
+      case tp: LazyRef =>
+        if !seen.contains(tp) then
+          seen += tp
+          traverse(tp.ref)
+      case _ => traverseChildren(t)
+  end Adjuster
+
+  /** Adjust dependencies to account for the delta of previous entry `prevEntry`
+   *  and the new bound `entry` for the type parameter `srcParam`.
+   */
+  def adjustDeps(entry: Type | Null, prevEntry: Type | Null, srcParam: TypeParamRef)(using Context): this.type =
+    val adjuster = new Adjuster(srcParam)
```

Comment:
Because we reuse the same `Adjuster` instance for multiple traversals, `seen` might accumulate some entries in a call that might influence the next call. I would just create a new adjuster instance each time to avoid worrying about that.

Reply:
I am going to `clear` the `seen` set instead. I am worried about creating too many adjusters otherwise.